desc/objectives/_neoclassical.py

```diff
@@ -0,0 +1,528 @@
+    num_pitch : int
+        Resolution for quadrature over velocity coordinate. Default is 50.
+    batch_size : int
+        Number of pitch values with which to compute simultaneously.
+        If given ``None``, then ``batch_size`` defaults to ``num_pitch``.
+
+    """
+
+    __doc__ = __doc__.rstrip() + collect_docs(
+        target_default="``target=0``.",
+        bounds_default="``target=0``.",
+        normalize_detail=" Note: Has no effect for this objective.",
+        normalize_target_detail=" Note: Has no effect for this objective.",
+        overwrite=_bounce_overwrite,
+    )
+
+    _coordinates = "r"
+    _units = "~"
+    _print_value_fmt = "Effective ripple ε: "
+
+    def __init__(
+        self,
+        eq,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        loss_function=None,
+        deriv_mode="fwd",
+        grid=None,
+        name="Effective ripple",
+        jac_chunk_size=None,
+        *,
+        X=16,  # X is cheap to increase.
+        Y=32,
+        # Y_B is expensive to increase if one does not fix num well per transit.
+        Y_B=None,
+        num_transit=20,
+        num_well=None,
+        num_quad=32,
+        num_pitch=50,
+        batch_size=None,
+    ):
+        if target is None and bounds is None:
+            target = 0.0
+
+        self._grid = grid
+        self._constants = {"quad_weights": 1}
+        self._X = X
+        self._Y = Y
+        Y_B = setdefault(Y_B, 2 * Y)
+        self._hyperparam = {
+            "Y_B": Y_B,
+            "num_transit": num_transit,
+            "num_well": setdefault(num_well, Y_B * num_transit),
+            "num_quad": num_quad,
+            "num_pitch": num_pitch,
+            "batch_size": batch_size,
+        }
+        if deriv_mode == "rev" and jac_chunk_size is None:
+            # Reverse mode is bottlenecked by coordinate mapping.
+            # Compute Jacobian one flux surface at a time.
+            jac_chunk_size = 1
+
+        super().__init__(
+            things=eq,
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            name=name,
+            jac_chunk_size=jac_chunk_size,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        eq = self.things[0]
+        if self._grid is None:
+            self._grid = LinearGrid(M=eq.M_grid, N=eq.N_grid, NFP=eq.NFP, sym=False)
+        assert self._grid.can_fft
+        # Constants can only hold 1D arrays for some reason.
```

Comment:
Does the error still persist if you make a copy instead of flattening it? Maybe it has to do something with pointers? It could be that somewhere in the code `FourierChebyshevSeries.nodes` changes size and it causes an inconsistency... I didn't look at it in detail but `.ravel()` creates a copy, just a thought

Reply:
Thanks. I should emphasize the issue has been "resolved" here, and I have made the GitHub issue so that we document that `constants` must be a 1D array only, or fix that limitation.

> In my case, I obtain the same error if I replace the objective.compute here with a dummy return jnp.ones(self._dim_f), so this error has nothing to do with bounce integrals.

With that mentioned modification, the fourierchebyshev.nodes array isn't referenced after it is made in `build`, so I don't think it's related.